File: src/email_parser/workflow/nodes/retrieve_profile.py
#!/usr/bin/env python3
"""
Retrieve Profile Node

Retrieves existing taxonomy profile from memory and applies
temporal decay to confidence scores.
"""

import logging
from datetime import datetime
from typing import Dict, Any

from ..state import WorkflowState, add_error
from ...memory.manager import MemoryManager
from ...memory.confidence import apply_temporal_decay, calculate_days_since_validation

logger = logging.getLogger(__name__)


def retrieve_existing_profile_node(
    state: WorkflowState,
    memory_manager: MemoryManager
) -> WorkflowState:
    """
    Retrieve existing profile from memory with temporal decay applied.

    Workflow Steps:
    1. Query all semantic memories for user
    2. Group memories by taxonomy section
    3. Apply temporal decay to confidence scores
    4. Structure profile by sections
    5. Update state with existing_profile

    Args:
        state: Current workflow state with user_id
        memory_manager: MemoryManager instance

    Returns:
        Updated state with existing_profile populated

    Example:
        >>> state = retrieve_existing_profile_node(state, memory_manager)
        >>> demographics = state['existing_profile']['demographics']
        >>> for mem in demographics:
        ...     print(f"{mem['value']}: {mem['confidence']:.2f}")
    """
    try:
        logger.info(f"Retrieving existing profile for user: {state['user_id']}")

        # Get all semantic memories
        all_memories = memory_manager.get_all_semantic_memories()
        logger.info(f"Retrieved {len(all_memories)} memories from store")

        # Apply temporal decay to each memory
        decayed_memories = []
        for memory in all_memories:
            # Calculate days since validation
            days_since = calculate_days_since_validation(memory["last_validated"])

            # Apply decay
            original_conf = memory["confidence"]
            decayed_conf = apply_temporal_decay(original_conf, days_since)

            # Update memory with decayed confidence
            memory_with_decay = memory.copy()
            memory_with_decay["confidence"] = decayed_conf
            memory_with_decay["days_since_validation"] = days_since

            if decayed_conf < original_conf:
                logger.debug(
                    f"Applied decay to {memory['memory_id']}: "
                    f"{original_conf:.3f} → {decayed_conf:.3f} "
                    f"({days_since} days)"
                )

            decayed_memories.append(memory_with_decay)

        # Group by section
        profile = {
            "demographics": [],
            "household": [],
            "interests": [],
            "purchase_intent": [],
            "actual_purchases": []
        }

        for memory in decayed_memories:
            section = memory.get("section", "unknown")
            if section in profile:
                profile[section].append(memory)
            else:
                logger.warning(f"Unknown section: {section}")

        # Log profile summary
        logger.info(f"Profile structure:")
        for section, memories in profile.items():
            if memories:
                avg_conf = sum(m["confidence"] for m in memories) / len(memories)
                logger.info(f"  {section}: {len(memories)} memories (avg confidence: {avg_conf:.3f})")

        # Update state
        state["existing_profile"] = profile

        logger.info("Profile retrieval complete")
        return state

    except Exception as e:
        logger.error(f"Error retrieving profile: {e}", exc_info=True)
        state = add_error(state, f"Failed to retrieve profile: {str(e)}")
        # Return empty profile on error
        state["existing_profile"] = {
            "demographics": [],
            "household": [],
            "interests": [],
            "purchase_intent": [],
            "actual_purchases": []
        }
        return state


def get_profile_summary(state: WorkflowState) -> Dict[str, Any]:
    """
    Get summary statistics of existing profile.

    Args:
        state: Workflow state with existing_profile

    Returns:
        Dictionary with profile statistics

    Example:
        >>> summary = get_profile_summary(state)
        >>> print(f"Total classifications: {summary['total_count']}")
        >>> print(f"Average confidence: {summary['average_confidence']:.2f}")
    """
    profile = state.get("existing_profile", {})

    total_count = 0
    total_confidence = 0.0
    high_confidence_count = 0
    stale_count = 0

    for section, memories in profile.items():
        for memory in memories:
            total_count += 1
            conf = memory.get("confidence", 0.0)
            total_confidence += conf

            if conf >= 0.8:
                high_confidence_count += 1

            if memory.get("days_since_validation", 0) >= 30:
                stale_count += 1

    avg_confidence = total_confidence / total_count if total_count > 0 else 0.0

    return {
        "total_count": total_count,
        "average_confidence": avg_confidence,
        "high_confidence_count": high_confidence_count,
        "stale_count": stale_count,
        "sections": {
            section: len(memories)
            for section, memories in profile.items()
            if memories
        }
    }

File: src/browser/agents/iab-classifier/nodes/retrieveProfile.ts
/**
 * Retrieve Profile Node
 *
 * 1:1 Port of Python retrieve_existing_profile_node from:
 * src/email_parser/workflow/nodes/retrieve_profile.py (lines 20-118)
 *
 * Retrieves existing taxonomy profile from memory and applies
 * temporal decay to confidence scores.
 */

import { WorkflowState } from '../state'
import { IndexedDBStore } from '@browser/store'
import { applyTemporalDecay, calculateDaysSinceValidation } from '../confidence'

/**
 * Retrieve existing profile from memory with temporal decay applied
 *
 * Python source: retrieve_profile.py:20-118 (retrieve_existing_profile_node)
 *
 * Workflow Steps:
 * 1. Query all semantic memories for user
 * 2. Group memories by taxonomy section
 * 3. Apply temporal decay to confidence scores
 * 4. Structure profile by sections
 * 5. Update state with existing_profile
 *
 * @param state Current workflow state with user_id
 * @param store IndexedDBStore instance
 * @returns Updated state with existing_profile populated
 */
export async function retrieveExistingProfileNode(
  state: typeof WorkflowState.State,
  store: IndexedDBStore
): Promise<Partial<typeof WorkflowState.State>> {
  try {
    console.info(`📂 Retrieving existing profile for user: ${state.user_id}`)

    // Python lines 50-52: Get all semantic memories
    // TODO: Implement store.getAllSemanticMemories()
    // For now, return empty profile
    const all_memories: Array<Record<string, any>> = []
    console.info(`Retrieved ${all_memories.length} memories from store`)

    // Python lines 54-76: Apply temporal decay to each memory
    const decayed_memories: Array<Record<string, any>> = []
    for (const memory of all_memories) {
      // Python line 58: Calculate days since validation
      const days_since = calculateDaysSinceValidation(memory.last_validated)

      // Python lines 61-62: Apply decay
      const original_conf = memory.confidence
      const decayed_conf = applyTemporalDecay(original_conf, days_since)

      // Python lines 64-67: Update memory with decayed confidence
      const memory_with_decay = {
        ...memory,
        confidence: decayed_conf,
        days_since_validation: days_since,
      }

      // Python lines 69-74: Log decay if applied
      if (decayed_conf < original_conf) {
        console.debug(
          `Applied decay to ${memory.memory_id}: ` +
          `${original_conf.toFixed(3)} → ${decayed_conf.toFixed(3)} ` +
          `(${days_since} days)`
        )
      }

      decayed_memories.push(memory_with_decay)
    }

    // Python lines 78-92: Group by section
    const profile: Record<string, Array<Record<string, any>>> = {
      demographics: [],
      household: [],
      interests: [],
      purchase_intent: [],
      actual_purchases: [],
    }

    for (const memory of decayed_memories) {
      const section = memory.section || 'unknown'
      if (section in profile) {
        profile[section].push(memory)
      } else {
        console.warn(`Unknown section: ${section}`)
      }
    }

    // Python lines 94-99: Log profile summary
    console.info('Profile structure:')
    for (const [section, memories] of Object.entries(profile)) {
      if (memories.length > 0) {
        const avg_conf = memories.reduce((sum, m) => sum + m.confidence, 0) / memories.length
        console.info(`  ${section}: ${memories.length} memories (avg confidence: ${avg_conf.toFixed(3)})`)
      }
    }

    // Python lines 101-105: Update state
    console.info('Profile retrieval complete')

    return {
      existing_profile: profile,
    }

  } catch (error) {
    // Python lines 107-118: Error handling
    console.error(`Error retrieving profile: ${error}`)

    const error_message = error instanceof Error ? error.message : String(error)

    // Python lines 110-117: Return empty profile on error
    return {
      existing_profile: {
        demographics: [],
        household: [],
        interests: [],
        purchase_intent: [],
        actual_purchases: [],
      },
      errors: [...(state.errors || []), `Failed to retrieve profile: ${error_message}`],
    }
  }
}
